src/controllers/collisions.rs

```diff
@@ -62,20 +74,77 @@ impl CollisionsController {
         state.score += SCORE_PER_ENEMY * killed_enemies;
     }
 
+    /// Handles collisions between the player and powerups
+    fn handle_powerup_collisions(state: &mut GameState, resources: &Resources) -> bool {
+        let mut gained_powerup = false;
+        let player = &mut state.world.player;
+        let powerups = &mut state.world.powerups;
+
+        if !player.is_dead {
+            if let Some((index, kind)) = powerups.iter().enumerate()
+                .find(|&(_, powerup)| powerup.collides_with(player))
+                .map(|(index, powerup)| (index, powerup.kind)) {
+                
+                gained_powerup = true;
+                
+                // Set player's powerup kind to the powerup we just picked up
+                player.powerup = Some(kind);
+                powerups.remove(index);
+
+                // Play the powerup sound
+                let _ = resources.powerup_sound.play();
+            }
+        }
+
+        return gained_powerup;
+    }
+
     /// Handles collisions between the player and the enemies
+    /// This function will return true if the player died
     fn handle_player_collisions(state: &mut GameState, resources: &Resources) -> bool {
-        let player_alive = !state.world.player.is_dead;
-        if player_alive && state.world.enemies.iter().any(|enemy| state.world.player.collides_with(enemy)) {
-            // Make an explosion where the player was
-            let ppos = state.world.player.position();
-            util::make_explosion(&mut state.world.particles, &ppos, 16);
-            // Mark the player as dead (to stop drawing it on screen)
-            state.world.player.is_dead = true;
-            // Play player_destroyed sound
-            let _ = resources.player_destroyed_sound.play();
-            return true;
+        let mut player_died = false;
+        let player = &mut state.world.player;
+
+        if !player.is_dead && state.world.enemies.iter().any(|enemy| player.collides_with(enemy)) {
+            // Remove shield powerup from player, also killing any enemies within close range
+            let mut had_shield = false;
+            if let Some(powerup) = player.powerup {
```

Comment:
I think you could use `if let Some(PowerupKind::Shield) = player.powerup` here. That way you get rid of the if below.

Reply:
DId not know you could unwrap enums like that - fantastic!